internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -521,8 +522,15 @@ private[inc] abstract class IncrementalCommon(
         invalidateClassesExternally(previous, externalAPIChange, isScalaSource)
       }.toSet
     }
+    val byMacroUsage = {
+      for {
+        class1 <- invalidatedClasses
+        class1Client <- previous.memberRef.internal.reverse(class1)
+        if previousAnalysis.apis.internalAPI(class1Client).hasMacro
```

Comment:
What is the dependency is transitive?

For example, in `macro-use` test case, if we create a new file `Data2.scala`:

```scala
class Data2 extends Data {}
```

And in `Macro.scala`, replace

```scala
val suffix = new Data().suffix
```

To 
```scala
val suffix = new Data2().suffix
```

Then `zincScripted/Test/run macros/macro-use` fails

Reply:
I couldn't tell why I wasn't reproducing your finding here, so I pushed the change as a commit to see whom CI corroborates with.

Reply:
What...

My sincere apology, now I also cannot reproduce the finding.

Maybe I ran the "counterexample" on wrong branch or something...

Reply:
My sincere apology again. My first comment was incorrect. Another level of indirection is needed.

Here's the test case that causes a fail.

[macro-use.zip](https://github.com/sbt/zinc/files/13417129/macro-use.zip)

Here's the test log on my machine

[scripted-macros-macro-use.log-1.log](https://github.com/sbt/zinc/files/13417133/scripted-macros-macro-use.log-1.log)

Reply:
Try

```scala
# Data2.scala
package lib

class Data2 {
  def suffix = new Data().suffix
}

val suffix = new Data2().suffix

# Macro.scala
val suffix = new Data2().suffix
```

Instead.

Here's a zip of the test case and test log on my local Machine.
[macro-use.zip](https://github.com/sbt/zinc/files/13418729/macro-use.zip)

[scripted-macros-macro-use.log-1.log](https://github.com/sbt/zinc/files/13418736/scripted-macros-macro-use.log-1.log)

Reply:
Also, while I cannot immediately construct a counter-example, would it be safer to place the check at end of every cycle instead of just start of the first cycle?

While doing the check for every cycle might be a massive performance hit, I have a concern that new Macro dependencies might be introduced in later cycles, such that the initial check won't catch them.

Reply:
Done